`infinispan-client` and `infinispan-cache` IT tests fail with GraalVM CE 25.2.4-dev

### Describe the bug

$title

### Expected behavior

Tests should pass

### Actual behavior

Tests fail with:

```
2026-06-23 01:33:24,268 WARN  [io.netty.channel.ChannelInitializer] (vert.x-eventloop-thread-0) Failed to initialize a channel. Closing: [id: 0x0d334412]: org.graalvm.nativeimage.MissingReflectionRegistrationError: Cannot reflectively invoke constructor 'public org.wildfly.security.sasl.digest.DigestClientFactory(java.security.Provider)'. To allow this operation, add the following to the 'reflection' section of 'reachability-metadata.json' and rebuild the native image:

    {
      "type": "org.wildfly.security.sasl.digest.DigestClientFactory",
      "methods": [
        {
          "name": "<init>",
          "parameterTypes": [
            "java.security.Provider"
          ]
        }
      ]
    }

The 'reachability-metadata.json' file should be located in 'META-INF/native-image/<group-id>/<artifact-id>/' of your project. For further help, see https://www.graalvm.org/latest/reference-manual/native-image/metadata/#reflection
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.MissingReflectionRegistrationUtils.reportInvokedExecutable(MissingReflectionRegistrationUtils.java:110)
Process exited with non-zero status: 1
Failed to launch the application. The application logs can be found at: /home/runner/work/mandrel/mandrel/quarkus/integration-tests/infinispan-client/target/quarkus.log
Last 50 line(s) of quarkus.log:
    	at io.netty.channel.AbstractChannel$AbstractUnsafe.ensureOpen(ChannelPromise)(Unknown Source)
	at java.base@25.0.3/java.lang.reflect.Constructor.acquireConstructorAccessor(Constructor.java:101)

    
    2026-06-23 01:33:24,287 runnervm7b5n9 /home/runner/work/mandrel/mandrel/quarkus/integration-tests/infinispan-client/target/quarkus-integration-test-infinispan-client-999-SNAPSHOT-runner[17730] ERROR [io.qua.run.Application] (main) Failed to start application: java.lang.RuntimeException: Failed to start quarkus
	at java.base@25.0.3/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:496)
    	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at java.base@25.0.3/java.lang.reflect.Constructor.newInstance(Constructor.java:483)
	at org.wildfly.security.WildFlyElytronBaseProvider$ProviderService.newInstance(WildFlyElytronBaseProvider.java:204)
	at org.wildfly.security.WildFlyElytronBaseProvider$ProviderService.newInstance(WildFlyElytronBaseProvider.java:187)
	at org.infinispan.commons.util.SaslUtils.findFactories(SaslUtils.java:72)
	at org.infinispan.commons.util.SaslUtils.getFactories(SaslUtils.java:55)
	at org.infinispan.commons.util.SaslUtils.getSaslClientFactories(SaslUtils.java:49)
	at org.infinispan.client.hotrod.impl.transport.netty.ChannelInitializer.getSaslClientFactory(ChannelInitializer.java:155)
	at org.infinispan.client.hotrod.impl.transport.netty.ChannelInitializer.initAuthentication(ChannelInitializer.java:132)
	at org.infinispan.client.hotrod.impl.transport.netty.ChannelInitializer.initChannel(ChannelInitializer.java:97)
	at io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:129)
	at io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:112)
	at io.netty.channel.AbstractChannelHandlerContext.callHandlerAdded(AbstractChannelHandlerContext.java:1130)
	at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:558)
	at io.netty.channel.DefaultChannelPipeline.access$100(DefaultChannelPipeline.java:45)
	at io.netty.channel.DefaultChannelPipeline$PendingHandlerAddedTask.execute(DefaultChannelPipeline.java:1410)
	at io.netty.channel.DefaultChannelPipeline.callHandlerAddedForAllHandlers(DefaultChannelPipeline.java:1064)
	at io.netty.channel.DefaultChannelPipeline.invokeHandlerAddedIfNeeded(DefaultChannelPipeline.java:599)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:513)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:428)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:485)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
    	at io.quarkus.runtime.Application.start(Application.java:112)
    	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:129)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
    	at io.quarkus.runtime.Quarkus.run(Quarkus.java:79)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
    	at io.quarkus.runtime.Quarkus.run(Quarkus.java:50)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
    	at io.quarkus.runtime.Quarkus.run(Quarkus.java:143)
    	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
    Caused by: jakarta.enterprise.inject.CreationException: Error creating synthetic bean [8GmzjuuTBXgKqMn01yfR-vdZKrE]: org.infinispan.client.hotrod.exceptions.TransportException:: io.netty.channel.StacklessClosedChannelException
    	at org.infinispan.client.hotrod.RemoteCacheManager_8GmzjuuTBXgKqMn01yfR-vdZKrE_Synthetic_Bean.create(Unknown Source)
    	at org.infinispan.client.hotrod.RemoteCacheManager_8GmzjuuTBXgKqMn01yfR-vdZKrE_Synthetic_Bean.create(Unknown Source)
    	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:119)
    	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:38)
    	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:35)
    	at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.ch4(Unknown Source)
    	at io.quarkus.arc.generator.Default_jakarta_enterprise_context_ApplicationScoped_ContextInstances.computeIfAbsent(Unknown Source)
    	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:35)
    	at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:23)
    	at org.infinispan.client.hotrod.RemoteCacheManager_8GmzjuuTBXgKqMn01yfR-vdZKrE_Synthetic_ClientProxy.arc$delegate(Unknown Source)
    	at org.infinispan.client.hotrod.RemoteCacheManager_8GmzjuuTBXgKqMn01yfR-vdZKrE_Synthetic_ClientProxy.getConfiguration(Unknown Source)
    	at io.quarkus.infinispan.client.runtime.InfinispanRecorder.initializeClient(InfinispanRecorder.java:79)
    	at io.quarkus.runner.recorded.InfinispanClientProcessor$infinispanClients405495887.deploy_0(Unknown Source)
    	at io.quarkus.runner.recorded.InfinispanClientProcessor$infinispanClients405495887.deploy(Unknown Source)
    	... 7 more
    Caused by: org.infinispan.client.hotrod.exceptions.TransportException:: io.netty.channel.StacklessClosedChannelException
    	at java.base@25.0.3/java.util.concurrent.CompletableFuture.wrapInExecutionException(CompletableFuture.java:345)
    	at java.base@25.0.3/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:440)
    	at java.base@25.0.3/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2117)
    	at org.infinispan.client.hotrod.impl.Util.await(Util.java:55)
    	at org.infinispan.client.hotrod.impl.Util.await(Util.java:50)
    	at org.infinispan.client.hotrod.impl.transport.netty.OperationDispatcher.await(OperationDispatcher.java:998)
    	at org.infinispan.client.hotrod.RemoteCacheManager.createRemoteCache(RemoteCacheManager.java:542)
    	at org.infinispan.client.hotrod.RemoteCacheManager.getCache(RemoteCacheManager.java:273)
    	at io.quarkus.infinispan.client.runtime.InfinispanClientProducer.registerSchemaInServer(InfinispanClientProducer.java:72)
    	at io.quarkus.infinispan.client.runtime.InfinispanClientProducer$1.apply(InfinispanClientProducer.java:502)
    	at io.quarkus.infinispan.client.runtime.InfinispanClientProducer$1.apply(InfinispanClientProducer.java:482)
    	at java.base@25.0.3/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1724)
    	at io.quarkus.infinispan.client.runtime.InfinispanClientProducer.getNamedRemoteCacheManager(InfinispanClientProducer.java:482)
    	at io.quarkus.infinispan.client.runtime.InfinispanRecorder$2.apply(InfinispanRecorder.java:44)
    	at io.quarkus.infinispan.client.runtime.InfinispanRecorder$2.apply(InfinispanRecorder.java:41)
    	at io.quarkus.infinispan.client.runtime.InfinispanRecorder$InfinispanClientSupplier.get(InfinispanRecorder.java:102)
    	at io.quarkus.arc.runtime.ArcRecorder$4.apply(ArcRecorder.java:138)
    	at io.quarkus.arc.runtime.ArcRecorder$4.apply(ArcRecorder.java:135)
    	at org.infinispan.client.hotrod.RemoteCacheManager_8GmzjuuTBXgKqMn01yfR-vdZKrE_Synthetic_Bean.createSynthetic(Unknown Source)
    	... 21 more
    Caused by: io.netty.channel.StacklessClosedChannelException
    	at io.netty.channel.AbstractChannel$AbstractUnsafe.ensureOpen(ChannelPromise)(Unknown Source)
    
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@25.0.3/java.lang.Thread.runWith(Thread.java:1487)
	at java.base@25.0.3/java.lang.Thread.run(Thread.java:1474)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:915)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:898)
```

See https://github.com/graalvm/mandrel/actions/runs/27995288543/job/82857532314https://github.com/graalvm/mandrel/actions/runs/27995288543/job/82857532314

### How to Reproduce?

1. Get GraalVM CE 25.2.4-dev build from https://github.com/graalvm/graalvm-ce-dev-builds/releases/tag/25.2.4-dev-20260623_0113
2. Decompress and set `$GRAALVM_HOME` path to point to the decompressed directory.
3. 
   ```
   ./mvnw clean verify -pl integration-tests/infinispan-client -Dnative -Dtest-containers -Dstart-containers
   ```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Mandrel or GraalVM version (if different from Java)

GraalVM CE 25.2.4-dev+9.1

### Quarkus version or git rev

f321ea32181bd3bd25a874e80418e701f75e7b1d

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageReflectConfigStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/NativeImageReflectConfigStep.java
@@ -3,6 +3,7 @@ package io.quarkus.deployment.steps;
 import java.io.IOException;
 import java.io.StringWriter;
 import java.nio.charset.StandardCharsets;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.List;
@@ -53,9 +54,11 @@ public class NativeImageReflectConfigStep {
 
         for (ServiceProviderBuildItem i : serviceProviderBuildItems) {
             for (String provider : i.providers()) {
-                // Register the nullary constructor
-                addReflectiveMethod(reflectiveClasses,
-                        new ReflectiveMethodBuildItem("Class registered as provider", provider, "<init>", new String[0]));
+                // Register all constructors as security providers may use
+                // java.security.Provider.Service.newInstance(Object constructorParameter) to instantiate the service
+                // through a different, than the nullary, constructor.
+                addReflectiveClass(reflectiveClasses, Collections.emptySet(),
+                        ReflectiveClassBuildItem.builder(provider).reason("Class registered as provider").build());
                 // Register public provider() method for lookkup to avoid throwing a MissingReflectionRegistrationError at run time.
                 // See ServiceLoader#loadProvider and ServiceLoader#findStaticProviderMethod.
                 addReflectiveMethod(reflectiveClasses,

